// scsslint-loader/dist/scsslint-loader.js
'use strict';

var _child_process = require('child_process');

var _scsslintStylish = require('./scsslint-stylish');

var _scsslintStylish2 = _interopRequireDefault(_scsslintStylish);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = function scsslintLoader(source, map) {
    var resourcePath = void 0;
    var args = [];
    var scsslintConfig = [];
    var webpack = [];
    var _cwd = process.cwd();
    /* eslint-disable no-unused-vars */
    function linterResults(_results, code) {
        /* eslint-enable no-unused-vars */
        var results = JSON.parse(_results);
        var res = [{
            filePath: resourcePath,
            messages: Object.assign([], results[resourcePath])
        }];

        if (res[0].messages.length > 0) {
            (function () {
                var warnings = 0;
                var errors = 0;
                var messages = res[0].messages;
                messages.forEach(function (linter) {
                    if (linter.severity === 'warning') {
                        warnings++;
                    } else if (linter.severity === 'error') {
                        errors++;
                    }
                });
                var msg = (0, _scsslintStylish2.default)(res);
                var emitter = void 0;
                if (errors > 0) {
                    emitter = webpack.emitError;
                } else {
                    emitter = webpack.emitWarning;
                }
                emitter(msg);
            })();
        }
    }

    function runLinter() {
        (0, _child_process.exec)(args.join(' '), {
            cwd: _cwd
        }, function (error, results, code) {
            if (error) {
                // console.log('scsslint error: ', error);
            }
            linterResults(results, code);
        });
    }

    function configLinter() {
        args.push('scss-lint');
        if (scsslintConfig.config) {
            args.push('-c');
            args.push(scsslintConfig.config);
        }
        args.push('--format=JSON');
        args = args.concat(resourcePath);
        args.push(';');
        runLinter();
    }

    var scsslintDefaults = {
        config: '.scss-lint.yml'
    };
    var globalSettings = this.options.scsslint || {};
    // const loader_settings = loaderUtils.parseQuery(this.query) || {};
    scsslintConfig = Object.assign(scsslintDefaults, globalSettings);

    this.cacheable();
    webpack = this;
    resourcePath = webpack.resourcePath;
    if (resourcePath.indexOf(_cwd) === 0) {
        resourcePath = resourcePath.substr(_cwd.length + 1);
    }
    configLinter();
    this.callback(null, source, map);
};
